Add documentation for Firebolt adapter (#941)
dbt-labs/docs.getdbt.com: website/docs/reference/resource-configs/firebolt-configs.md

---
title: "Firebolt configurations"
id: "firebolt-configs"
---

> ### Setup Recommendations

Reviewer: I don't think this information belongs on the configs doc. The configs doc should be purely model configs. I'd remove this section altogether and just leave it in the adapter repo's README.

Author: This _is_ a config file, and these _are_ configuration suggestions. At the same time, it seems kind of granular. @swanderz and I had a discussion  yesterday about what to include here vs. the readme. I think you two should discuss this before I go removing anything.

Reviewer: In dbt docs, this page will be listed as a sub-page of "Adapter-specific configs". `quote_columns` isn't adapter specific.

Reviewer 2: I personally think this would be good to leave in the adapter-specific config because you can give a warning as to why the warning is happening and how to fix it (especially because we don't have adapter specific troubleshooting pages)

Reviewer: Okay, that's fine with me. Let's just leave the quote_columns thing in there then.

```sql
-- macros/generate_alias_name.sql
{% macro generate_alias_name(custom_alias_name=none, node=none) -%}
    {%- if custom_alias_name is none -%}
        {{ node.schema }}__{{ node.name }}
    {%- else -%}
        {{ node.schema }}__{{ custom_alias_name | trim }}
    {%- endif -%}
{%- endmacro %}
```

For an example of how this works, let’s say Shahar and Eric are both working on the same project.

In her `.dbt/profiles.yml`, Sharar sets `schema=sh`, whereas Eric sets `schema=er` in his. When each runs the `customers` model, the models will land in the database as tables named `sh_customers` and `er_customers`, respectively. When running dbt in production, you would use yet another `profiles.yml` with nothing set in the `schema` field (or any string of your choice).

> ## Feature Support

Reviewer: This information also does not belong here - let's remove.

The table below shows which dbt and Firebolt features are supported by the adapter. dbt-firebolt is under active development and will be gradually unlocking more features over time.

| Feature                      | Supported          |
|------------------------------|--------------------|
| Table materializations       | :white_check_mark: |
| Ephemeral materializations   | :white_check_mark: |
| View materializations        | :white_check_mark: |
| Incremental materializations | :x: |
| Seeds                        | :white_check_mark: |
| Tests                        | :white_check_mark: |
| Documentation                | :white_check_mark: |
| Snapshots                    | :x: |
| Custom schemas               | :x: (see [workaround](https://github.com/firebolt-db/dbt-firebolt#supporting-concurrent-development)) |
| Custom databases             | :x: |
| Source freshness             | :white_check_mark: |
| External tables              | :white_check_mark: |
| Primary indexes              | :white_check_mark: |
| Aggregating indexes          | :white_check_mark: |
| Join indexes                 | :white_check_mark: |
| Partitioned tables           | :x: |

> 

Reviewer 3: ```suggestion
```

## Using dbt-firebolt

> ### Model Configuration for Fact Tables

Reviewer: keep this section

A dbt model can be created as a Firebolt fact table and configured using the following syntax:

> Project file syntax

Reviewer: We should mark this up such that the three code examples display as tabs in the same code snippet. See https://docs.getdbt.com/reference/model-configs for an example.

```yml
models:
  <resource-path>:
    +materialized: table
    +table_type: fact
    +primary_index: [ <column-name>, ... ]
    +indexes:
      - type: aggregating
        key_column: [ <column-name>, ... ]
        aggregation: [ <agg-sql>, ... ]
      ...
```

Property file syntax
```yml
models:
  - name: <model-name>
    config:
      materialized: table
      table_type: fact
      primary_index: [ <column-name>, ... ]
      indexes:
        - type: aggregating | join
          key_column: [ <column-name>, ... ]
          aggregation: [ <agg-sql>, ... ]
        ...
```

Config block syntax
```
{{ config(
    materialized = "table"
    table_type = "fact"
    primary_index = [ "<column-name>", ... ],
    indexes = [
      {
        type = "aggregating"
        key_column = [ "<column-name>", ... ],
        aggregation = [ "<agg-sql>", ... ],
      },
      ...
    ]
) }}

> ```

Reviewer 3: this should do it? the preview rendering is failing bc this is complicated so [here's a gist for it](https://gist.github.com/swanderz/01e1145e596801dfbd95bfba1a2c17da#file-excerpt_1-md)
```suggestion
<Tabs
  defaultValue="Project"
  values={[
    { label: 'Project file', value: 'Project'},
    { label: 'schema file', value:'Property'},
    { label: 'config block', value: 'Config'},
    ]
}>

<TabItem value="Project">
<File name='dbt_project.yml'>

```yml
models:
  <resource-path>:
    +materialized: table
    +table_type: fact
    +primary_index: [ <column-name>, ... ]
    +indexes:
      - type: aggregating
        key_column: [ <column-name>, ... ]
        aggregation: [ <agg-sql>, ... ]
      ...
```

</File>
</TabItem>

<TabItem value="Project">
<File name='schema.yml'>

```yml
models:
  - name: <model-name>
    config:
      materialized: table
      table_type: fact
      primary_index: [ <column-name>, ... ]
      indexes:
        - type: aggregating | join
          key_column: [ <column-name>, ... ]
          aggregation: [ <agg-sql>, ... ]
        ...
```

</File>
</TabItem>
<TabItem value="Config">
<File name='my_model.sql'>

```
{{ config(
    materialized = "table"
    table_type = "fact"
    primary_index = [ "<column-name>", ... ],
    indexes = [
      {
        type = "aggregating"
        key_column = [ "<column-name>", ... ],
        aggregation = [ "<agg-sql>", ... ],
      },
      ...
    ]
) }}
```

</File>

</TabItem>

</Tabs>
```

| `primary_index`   | Sets the primary index for the fact table using the inputted list of column names from the model. Required for fact tables. |
| `indexes`         | A list of aggregating indexes to create on the fact table. |
| `type`            | Specifies whether the index is an aggregating index or join index. Join indexes only apply to dimension tables, so for fact tables set to `aggregating`. |
| `key_column`      | Sets the grouping of the aggregating index using the inputted list of column names from the model. |
| `aggregation`     | Sets the aggregations on the aggregating index using the inputted list of SQL agg expressions. |

#### Example of a Fact Table With an Aggregating Index (Config Block Syntax)

```
{{ config(
    materialized = "table",
    table_type = "fact",
    primary_index = "id",
    indexes = [
      {
        type: "aggregating",
        key_column: "order_id",
        aggregation: ["COUNT(DISTINCT status)", "AVG(customer_id)"]
      }
    ]
) }}
```

### Model Configuration for Dimension Tables

A dbt model can be created as a Firebolt dimension table and configured using the following syntax:

> Project file syntax

Reviewer: See comment above about consolidating the three examples into one code block

> Project file syntax
> ```yml
> models:
>   <resource-path>:
>     +materialized: table
>     +table_type: dimension
>     +indexes:
>       - type: join
>         join_column: <column-name>
>         dimension_column: [ <column-name>, ... ]
>       ...
> ```
> 
> Property file syntax
> ```yml
> models:
>   - name: <model-name>
>     config:
>       materialized: table
>       table_type: dimension
>       indexes:
>         - type: join
>           join_column: <column-name>
>           dimension_column: [ <column-name>, ... ]
>         ...
> ```
> 
> Config block syntax
> ```
> {{ config(
>     materialized = "table"
>     table_type = "dimension",
>     indexes = [
>       {
>         type = "join",
>         join_column = "<column-name>",
>         dimension_column: [ "<column-name>", ... ]
>       },
>       ...
>     ]
> ) }}
> ```

Reviewer 3: [second gist](https://gist.github.com/swanderz/01e1145e596801dfbd95bfba1a2c17da#file-excerpt_2-md) 

```suggestion
<Tabs
  defaultValue="Project"
  values={[
    { label: 'Project file', value: 'Project'},
    { label: 'schema file', value:'Property'},
    { label: 'config block', value: 'Config'},
    ]
}>

<TabItem value="Project">
<File name='dbt_project.yml'>

```yml
models:
  <resource-path>:
    +materialized: table
    +table_type: dimension
    +indexes:
      - type: join
        join_column: <column-name>
        dimension_column: [ <column-name>, ... ]
      ...
```

</File>

</TabItem>

<TabItem value="Project">

<File name='schema.yml'>

```yml
models:
  - name: <model-name>
    config:
      materialized: table
      table_type: dimension
      indexes:
        - type: join
          join_column: <column-name>
          dimension_column: [ <column-name>, ... ]
        ...
```

</File>

</TabItem>

<TabItem value="Config">

<File name='my_model.sql'>

Config block syntax
```
{{ config(
    materialized = "table"
    table_type = "dimension",
    indexes = [
      {
        type = "join",
        join_column = "<column-name>",
        dimension_column: [ "<column-name>", ... ]
      },
      ...
    ]
) }}
```

</File>

</TabItem>

</Tabs>

```

| `materialized`     | How the model will be materialized into Firebolt. Must be `table` to create a dimension table. |
| `table_type`       | Whether the materialized table will be a [fact or dimension](https://docs.firebolt.io/concepts/working-with-tables#fact-and-dimension-tables) table. |
| `indexes`          | A list of join indexes to create on the dimension table. |
| `type`             | Specifies whether the index is an aggregating index or join index. Aggregating indexes only apply to fact tables, so for dimension tables set to `join`. |
| `join_column`      | Sets the join key of the join index using the inputted column name from the model. |
| `dimension_column` | Sets the columns to be loaded into memory on the join index using the inputted list of column names from the mode. |

#### Example of a Dimension Table With a Join Index (Config Block Syntax)

```
{{ config(
    materialized = "table",
    table_type = "dimension",
    indexes = [
      {
        type: "join",
        join_column: "order_id",
        dimension_column: ["customer_id", "status"]
      }
    ]
) }}
```

> ### How Aggregating Indexes and Join Indexes Are Named

Reviewer: Let's pull this out and leave it in the adapter repo README

In dbt-firebolt, you do not provide names for aggregating indexes and join indexes; they are named programmatically. dbt will generate index names using the following convention:
```
<table-name>__<key-column>__<index-type>_<unix-timestamp-at-execution>
```

For example, a join index could be named `my_users__id__join_1633504263` and an aggregating index could be named `my_orders__order_date__aggregating_1633504263`.

> ### Managing Ingestion via External Tables

Reviewer: Pull out and leave in README

In addition to specifying the columns, an external table may specify partitions. Partitions are not columns and they cannot have the same name as columns. To avoid yaml parsing errors, remember to encase string literals (such as the `url` and `object_pattern` values) in single quotation marks.

#### dbt_project.yml Syntax For an External Table

```yml
sources:
  - name: firebolt_external
    schema: "{{ target.schema }}"
    loader: S3

    tables:
      - name: <table-name>
        external:
          url: 's3://<bucket_name>/'
          object_pattern: '<regex>'
          type: '<type>'
          credentials:
            internal_role_arn: arn:aws:iam::id:<role>/<bucket-name>
            external_role_id: <external-id>
          object_pattern: '<regex>'
          compression: '<compression-type>'
          partitions:
            - name: <partition-name>
              data_type: <partition-type>
              regex: '<partition-definition-regex>'
          columns:
            - name: <column-name>
              data_type: <type>

> ```

Reviewer 3: ```suggestion
```

```sql
-- macros/generate_alias_name.sql
{% macro generate_alias_name(custom_alias_name=none, node=none) -%}
    {%- if custom_alias_name is none -%}
        {{ node.schema }}__{{ node.name }}
    {%- else -%}
        {{ node.schema }}__{{ custom_alias_name | trim }}
    {%- endif -%}
{%- endmacro %}
```

For an example of how this works, let’s say Shahar and Eric are both working on the same project.

In her `.dbt/profiles.yml`, Sharar sets `schema=sh`, whereas Eric sets `schema=er` in his. When each runs the `customers` model, the models will land in the database as tables named `sh_customers` and `er_customers`, respectively. When running dbt in production, you would use yet another `profiles.yml` with nothing set in the `schema` field (or any string of your choice).

## Feature Support

> The table below shows which dbt and Firebolt features are supported by the adapter. dbt-firebolt is under active development and will be gradually unlocking more features over time.

Reviewer 2: This section should be under the profile page. [See here Materialize for reference](https://docs.getdbt.com/reference/warehouse-profiles/materialize-profile)

Reviewer: I disagree; I think it is unexpected to find this information in a doc entitled "Firebolt Profile". The expectation is to find information only pertaining to creating profiles for Firebolt connections. My preference is to have it in the repo readme for dbt-firebolt and not in dbt docs at all.

---
title: "Firebolt configurations"
id: "firebolt-configs"
---

> ## Using dbt-firebolt

Reviewer: @ericf-firebolt This is an example of a header that is no longer relevant due to the change in context between the dbt docs vs our readme - please remove

### `quote_columns`

To prevent a warning, you should add a configuration as below to your `dbt_project.yml`. For more info, see the [relevant dbt docs page](https://docs.getdbt.com/reference/resource-configs/quote_columns).

<File name>

> ```yml

Reviewer: @ericf-firebolt This is not rendering properly on the rich diff in GH, which doesn't necessarily mean it won't render properly on dbt docs, but I just wanted to flag that. Might need some newlines.

> ### `quote_columns`
> 
> To prevent a warning, you should add a configuration as below to your `dbt_project.yml`. For more info, see the [relevant dbt docs page](https://docs.getdbt.com/reference/resource-configs/quote_columns).
> 
> 
> <File name>
> ```yml
> seeds:
>   +quote_columns: false  #or `true` if you have csv column headers with spaces
> ```
> </File>

Reviewer 3: do we still need this section on `quote_columns`? my vote is to strike it.

something looks weird w/ it right now anyway

cc: @kevinmarr @amychen1776 @ericf-firebolt 

<img width="685" alt="image" src="https://user-images.githubusercontent.com/8158673/145123522-c9af34e8-25e0-445c-9d3d-882146630f7c.png">

```suggestion
```

Reviewer 2: Since this in the profiles config page, I don't think so.

Author: The thing that was weird is that it said "yml" after the triple ticks and not "yaml".

### Model Configuration for Fact Tables

> A dbt model can be created as a Firebolt fact table and configured using the following syntax:

Reviewer 2: ```suggestion
A dbt model be materialized as a Firebolt fact table and configured using the following syntax:
```

<Tabs
  groupId="config-languages"
  defaultValue="project-yaml"
  values={[
    { label: 'Project file', value: 'project-yaml', },
    { label: 'Property file', value: 'property-yaml', },
    { label: 'Config block', value: 'config', },
  ]
}>

<TabItem value="project-yaml">
<File name='dbt_project.yml'>

```yaml
models:

>   [<resource-path>]:

Reviewer 2: Should these brackets be here (for resource-path)?

Reviewer: No - not sure how those got in because it is not displayed that way in our [readme](https://github.com/firebolt-db/dbt-firebolt).

    +materialized: table
    +table_type: fact
    +primary_index: [ <column-name>, ... ]
    +indexes:
      - type: aggregating
        key_column: [ <column-name>, ... ]
        aggregation: [ <agg-sql>, ... ]
      ...
```

</File>
</TabItem>

<TabItem value="property-yaml">

> <File name='models/properties.yml'>

Reviewer 2: Wait, this looks odd. @swanderz Users should only be applying model configurations on the model config block and the dbt_project.yml file. This is not a blocker for me but feels odd to recommend to a user from a workflow perspective (why have a model file with configs when the logic is easier for the user in the model file or project.yml for massive application)

Reviewer 3: @kevinmarr @ericf-firebolt just chatted w/ @amychen1776 on this, and the dbt perspective is that while yes, you can set model configs in property files, it really isn't recommended. Her advice is to strike the property file examples for both dimension and fact tables

Reviewer: Great, let's axe 'em.

| `table_type`      | Whether the materialized table will be a [fact or dimension](https://docs.firebolt.io/concepts/working-with-tables#fact-and-dimension-tables) table. |
| `primary_index`   | Sets the primary index for the fact table using the inputted list of column names from the model. Required for fact tables. |
| `indexes`         | A list of aggregating indexes to create on the fact table. |
| `type`            | Specifies whether the index is an aggregating index or join index. Join indexes only apply to dimension tables, so for fact tables set to `aggregating`. |
| `key_column`      | Sets the grouping of the aggregating index using the inputted list of column names from the model. |
| `aggregation`     | Sets the aggregations on the aggregating index using the inputted list of SQL agg expressions. |

#### Example of a Fact Table With an Aggregating Index (Config Block Syntax)

```
{{ config(
    materialized = "table",
    table_type = "fact",
    primary_index = "id",
    indexes = [
      {
        type: "aggregating",
        key_column: "order_id",
        aggregation: ["COUNT(DISTINCT status)", "AVG(customer_id)"]
      }
    ]
) }}
```

> ### Model Configuration for Dimension Tables

Reviewer: I realized that we left out `primary_index` in the dimension tables section. @ericf-firebolt Could you add it in? It will need to be added to the code examples as well as the table.

### Managing Ingestion via External Tables

`dbt-firebolt` supports dbt's [external tables feature](https://docs.getdbt.com/reference/resource-properties/external), which allows dbt to manage the table ingestion process from S3 into Firebolt. This is an optional feature but can be highly convenient depending on your use case.

More information on using external tables including properly configuring IAM can be found in the Firebolt [documentation](https://docs.firebolt.io/sql-reference/commands/ddl-commands#create-external-table).

#### Installation of External Tables Package

To install and use `dbt-external-tables` with Firebolt, you must:

1. Add this package to your packages.yml:

    ```yaml
    packages:
      - package: dbt-labs/dbt_external_tables
        version: <version>
    ```

2. Add these fields to your `dbt_project.yml`:

    ```yaml
    dispatch:
      - macro_namespace: dbt_external_tables
        search_order: ['dbt', 'dbt_external_tables']
    ```

3. Pull in the `packages.yml` dependencies by calling `dbt deps`.

> #### Using External Tables
> 
> To use external tables, you must define a table as `external` in your `dbt_project.yml` file. Every external table must contain the fields `url`, `type`, and `object_pattern`. Note that the Firebolt external table specification requires fewer fields than what is specified in the dbt documentation.
> 
> In addition to specifying the columns, an external table may specify partitions. Partitions are not columns and they cannot have the same name as columns. To avoid yaml parsing errors, remember to encase string literals (such as the `url` and `object_pattern` values) in single quotation marks.
> 
> #### dbt_project.yml Syntax For an External Table
> 
> ```yaml
> sources:
>   - name: firebolt_external
>     schema: "{{ target.schema }}"
>     loader: S3
> 
>     tables:
>       - name: <table-name>
>         external:
>           url: 's3://<bucket_name>/'
>           object_pattern: '<regex>'
>           type: '<type>'
>           credentials:
>             internal_role_arn: arn:aws:iam::id:<role>/<bucket-name>
>             external_role_id: <external-id>
>           object_pattern: '<regex>'
>           compression: '<compression-type>'
>           partitions:
>             - name: <partition-name>
>               data_type: <partition-type>
>               regex: '<partition-definition-regex>'
>           columns:
>             - name: <column-name>
>               data_type: <type>
> ```

Reviewer 3: we've already gutted most of the "using dbt-external tables" section, I vote we get rid of the rest of it?
```suggestion
```

---
title: "Firebolt configurations"
id: "firebolt-configs"
---

### Setup Recommendations

#### `quote_columns`

To prevent a warning, you should add a configuration as below to your `dbt_project.yml`. For more info, see the [relevant dbt docs page](https://docs.getdbt.com/reference/resource-configs/quote_columns).

```yaml
seeds:
  +quote_columns: false  #or `true` if you have csv column headers with spaces
```

> #### Supporting Concurrent Development

Reviewer: Let's move this to the profiles doc
